// @pulumi/aws/ecrpublic/repository.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Provides a Public Elastic Container Registry Repository.
 *
 * ## Import
 *
 * ECR Public Repositories can be imported using the `repository_name`, e.g.
 *
 * ```sh
 *  $ pulumi import aws:ecrpublic/repository:Repository example example
 * ```
 */
class Repository extends pulumi.CustomResource {
    constructor(name, argsOrState, opts) {
        let inputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            inputs["arn"] = state ? state.arn : undefined;
            inputs["catalogData"] = state ? state.catalogData : undefined;
            inputs["forceDestroy"] = state ? state.forceDestroy : undefined;
            inputs["registryId"] = state ? state.registryId : undefined;
            inputs["repositoryName"] = state ? state.repositoryName : undefined;
            inputs["repositoryUri"] = state ? state.repositoryUri : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.repositoryName === undefined) && !opts.urn) {
                throw new Error("Missing required property 'repositoryName'");
            }
            inputs["catalogData"] = args ? args.catalogData : undefined;
            inputs["forceDestroy"] = args ? args.forceDestroy : undefined;
            inputs["repositoryName"] = args ? args.repositoryName : undefined;
            inputs["arn"] = undefined /*out*/;
            inputs["registryId"] = undefined /*out*/;
            inputs["repositoryUri"] = undefined /*out*/;
        }
        if (!opts.version) {
            opts = pulumi.mergeOptions(opts, { version: utilities.getVersion() });
        }
        super(Repository.__pulumiType, name, inputs, opts);
    }
    /**
     * Get an existing Repository resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new Repository(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of Repository.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === Repository.__pulumiType;
    }
}
exports.Repository = Repository;
/** @internal */
Repository.__pulumiType = 'aws:ecrpublic/repository:Repository';
